play-json/shared/src/main/scala-3/play/api/libs/json/QuotesHelper.scala

```diff
@@ -43,51 +42,27 @@ private[json] trait QuotesHelper {
    * Class `Lorem` is listed through `SubFoo`,
    * but `SubFoo` itself is not returned.
    */
-  final def knownSubclasses(tpr: TypeRepr): Option[List[TypeRepr]] =
-    tpr.classSymbol.flatMap { cls =>
-      @annotation.tailrec
-      def subclasses(
-          children: List[Tree],
-          out: List[TypeRepr]
-      ): List[TypeRepr] = {
-        val childTpr = children.headOption.collect {
-          case tpd: Typed =>
-            tpd.tpt.tpe
-
-          case vd: ValDef =>
-            vd.tpt.tpe
-
-          case cd: ClassDef =>
-            cd.constructor.returnTpt.tpe
-
-        }
-
-        childTpr match {
-          case Some(child) => {
-            val tpeSym = child.typeSymbol
-
-            if (tpeSym.flags.is(Flags.Abstract) &&
-                tpeSym.flags.is(Flags.Sealed) &&
-                !(child <:< anyValTpe)) ||
-              (tpeSym.flags.is(Flags.Sealed) &&
-                tpeSym.flags.is(Flags.Trait))
-            then {
-              // Ignore sub-trait itself, but check the sub-sub-classes
-              subclasses(tpeSym.children.map(_.tree) ::: children.tail, out)
-            } else {
-              subclasses(children.tail, child :: out)
-            }
+  final def knownSubclasses(tpr: TypeRepr): Option[List[TypeRepr]] = {
+    def gatherNestedSubtypes[Parent: Type, Elems: Type](using Quotes): List[TypeRepr] =
+      Type.of[Elems] match {
+        case '[elem *: elems] =>
+          Expr.summon[Mirror.Of[elem]] match {
+            case Some('{ $sum: Mirror.SumOf[elem] { type MirroredElemTypes = elementTypes } }) =>
+              gatherNestedSubtypes[elem, elementTypes] ++ gatherNestedSubtypes[Parent, elems]
+            case _ =>
+              TypeRepr.of[elem] :: gatherNestedSubtypes[Parent, elems]
           }
-
-          case _ =>
-            out.reverse
-        }
+        case '[EmptyTuple] => Nil
       }
 
-      val types = subclasses(cls.children.map(_.tree), Nil)
-
-      if types.isEmpty then None else Some(types)
+    tpr.asType match {
+      case '[t] =>
+        Expr.summon[Mirror.Of[t]].collect {
```

Comment:
I think that a less WIP implementation of this should probably use `Implicits.search` so that it could surface any implicit search error messages to the user